### PR TITLE
Add random port finder methods to LocalPortChecker

### LINKED ISSUE #842: New methods in LocalPortChecker to find random open ports
### Summary

Add new methods to find _random_ ports to `LocalPortChecker`.

These should:

* find a random open port
* find a random open port starting from a specific port, inclusive
* find a random open port starting from a specific port, exclusive

### Proposed Methods

| Method Signature | Description |
| ------------- | ------------- |
| `OptionalInt findRandomOpenPort()`  | Finds a random port between 1 and 65535  |
| `OptionalInt findRandomOpenPortFrom(int port)`  | Finds a random port between `port` (inclusive) and 65535 |
| `OptionalInt findRandomOpenPortAbove(int port)` | Finds a random port between `port` (exclusive) and 65535 |

### Background

We currently only provide a way to find the first open port starting from a specific port (inclusive or exclusive). We do not provide a way to find a random port, which can be useful in certain situations.

One example is the need to launch a test server, for example a [CuratorTestingServerExtension](https://github.com/kiwiproject/kiwi-test/blob/master/src/main/java/org/kiwiproject/test/curator/CuratorTestingServerExtension.java) in tests. If there are several tests using this extension, and if tests are run concurrently, this can cause problems with ports being already bound. This happened to me yesterday while running `mvn test` on multiple repositories concurrently; I saw "address already in use" several times because tests using `CuratorTestingServerExtension` were all using the _same_ port.

By allowing random open port selection, the chance of port conflicts should be much less than if tests always start from 1024, or worse if they always uses the same port like in the tests mentioned above. Once these methods are in place, it would be useful to find and change the default behavior in our libraries, e.g. `CuratorTestingServerExtension`, to use random port selection, or at least to make it easy to use random port selection if we don't want to change the existing behavior.

### PR DESCRIPTION
**Pull request by sleberknight**
Closes #842

### CODE REVIEW

File: src/main/java/org/kiwiproject/net/LocalPortChecker.java [hunk 1]
```diff
@@ -65,10 +66,68 @@ public OptionalInt findFirstOpenPortFrom(int port) {
...
         return IntStream.rangeClosed(port + 1, MAX_PORT)
                 .filter(this::isPortAvailable)
                 .findFirst();
     }
+
+    /**
+     * Find a random open port. If no available port is found, return an empty {@link OptionalInt}.
+     *
+     * @return an optional containing the first open port, or an empty optional
+     * @throws IllegalArgumentException if port is not between 1 and {@link #MAX_PORT}
```
Comment (dsingley) [inline src/main/java/org/kiwiproject/net/LocalPortChecker.java:80]: `@throws` is incorrect here

File: src/main/java/org/kiwiproject/net/LocalPortChecker.java [hunk 2]
```diff
@@ -65,10 +66,68 @@ public OptionalInt findFirstOpenPortFrom(int port) {
...
+    /**
+     * Find a random open port starting at the given port. If no available port is found, return an empty
+     * {@link OptionalInt}.
+     *
+     * @param port the port to start from (inclusive)
+     * @return an optional containing the first open port, or an empty optional
+     * @implNote Since this method looks for availability of a specific port, zero is invalid because it is a reserved
+     * port used to indicate a random port and is never bound directly to.
+     */
+    public OptionalInt findRandomOpenPortFrom(int port) {
```
Comment (dsingley) [inline src/main/java/org/kiwiproject/net/LocalPortChecker.java:95]: `@throws` should be here

**Review by dsingley — CHANGES_REQUESTED** [review]

File: src/main/java/org/kiwiproject/net/LocalPortChecker.java (hunk 1 above)
Comment (sleberknight) [reply src/main/java/org/kiwiproject/net/LocalPortChecker.java:80]: Good catch

File: src/main/java/org/kiwiproject/net/LocalPortChecker.java (hunk 2 above)
Comment (sleberknight) [reply src/main/java/org/kiwiproject/net/LocalPortChecker.java:95]: Ditto (good catch)

**Review by dsingley — APPROVED** [review]

**Review by dsingley — APPROVED** [review]

**Pull request merged**